### PR TITLE
Add an optimizer rule to remove redundant assignments in ProjectNode

### LINKED ISSUE #14683: Compilation error when same lambda projection appears multiple times
The following query would fail with compilation error:
```
SELECT
    reduce(x, 0, (s, x) -> s + x, (s) -> s),
    reduce(x, 0, (s, x) -> s + x, (s) -> s)
FROM (
    VALUES (array[1,2,3]), array[3, 5, 6]) t(x)
```
This is because we do projection deduplication only in `Expression` and the way we generate `LambdaExpression`s, the two projections above would be converted to different `Expression`. However, to enable common sub-expression optimization for lambda, we introduced canonicalization of `LambdaDefinitionExpression`, which is the `RowExpression` correspondence of  `LambdaExpression` (#14524). As a result, after the translation, the above two projections would be the same.

### PR DESCRIPTION
**Pull request by rongrong**
Fixes #14683

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Context context)
+    {
+        Map<Boolean, List<Map.Entry<VariableReferenceExpression, RowExpression>>> projections = node.getAssignments().entrySet().stream()
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:48]: nit: static import `Map.Entry`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                .collect(Collectors.partitioningBy(entry -> entry.getValue() instanceof VariableReferenceExpression));
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:49]: static import `Collectors.partitioningBy`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        Map<RowExpression, ImmutableMap<VariableReferenceExpression, RowExpression>> uniqueProjections = projections.get(false).stream()
+                .collect(Collectors.groupingBy(Map.Entry::getValue, toImmutableMap(Map.Entry::getKey, Map.Entry::getValue)));
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:51]: static import `Collectors.partitioningBy`, Map.Entry

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if (uniqueProjections.size() == projections.get(false).size()) {
+            return Result.empty();
+        }
+        Assignments.Builder childAssignments = Assignments.builder();
+        Assignments.Builder parentAssignments = Assignments.builder();
+        projections.get(true).forEach(entry -> childAssignments.put(entry.getKey(), entry.getValue()));
+        projections.get(true).forEach(entry -> parentAssignments.put(entry.getKey(), entry.getKey()));
+        for (Map.Entry<RowExpression, ImmutableMap<VariableReferenceExpression, RowExpression>> entry : uniqueProjections.entrySet()) {
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:59]: static import Map.Entry
Also, I would normally only use interfaces here:
```
for (Entry<RowExpression, Map<VariableReferenceExpression, RowExpression>>)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Context context)
+    {
+        Map<Boolean, List<Map.Entry<VariableReferenceExpression, RowExpression>>> projections = node.getAssignments().entrySet().stream()
+                .collect(Collectors.partitioningBy(entry -> entry.getValue() instanceof VariableReferenceExpression));
+        Map<RowExpression, ImmutableMap<VariableReferenceExpression, RowExpression>> uniqueProjections = projections.get(false).stream()
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:50]: I would normally only use interfaces here:
```
for (Entry<RowExpression, Map<VariableReferenceExpression, RowExpression>>)
```

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 6]
```diff
@@ -8447,6 +8447,14 @@ public void testSetAgg()
                 "select count(distinct orderdate) from orders");
     }
 
+    @Test
+    public void testRedundantLambda()
+    {
+        assertQuery(
+                "SELECT x, reduce(x, 0, (s, x) -> s + x, s -> s), reduce(x, 0, (s, x) -> s + x, s -> s) FROM (VALUES (array[1, 2, 3])) t(x)",
+                "SELECT array[1, 2, 3], 6, 6");
+    }
```
Comment (caithagoras) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:8456]: I tried this query out on the master with the CSE enabled and it is succeeding. Is there a case which fails on the master but succeeds with this fix?

Comment (caithagoras) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:8456]: My mistake, the issue on master but is not in a release.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java (hunk 5 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:50]: compiler actually complains if I use Map here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:51]: Most call site of these are not static import if you search the codebase.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java (hunk 1 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneRedundantProjectionAssignments.java:48]: We normally don't static import `Map.Entry`.

**Pull request merged**